src/transformers/trainer_tf.py

```diff
@@ -70,20 +72,34 @@ def __init__(
     def get_train_tfdataset(self) -> tf.data.Dataset:
         if self.train_dataset is None:
             raise ValueError("Trainer: training requires a train_dataset.")
-
-        self.num_train_examples = self.train_dataset.reduce(tf.constant(0), lambda x, _: x + 1).numpy()
+        """
+        # self.num_train_examples = self.train_dataset.reduce(tf.constant(0), lambda x, _: x + 1).numpy()
+        self.num_train_examples = 0
+        for ex in self.train_dataset:
+            self.num_train_examples += 1
+        """
 
         if self.args.max_steps > 0:
             self.train_steps = self.args.max_steps
         else:
             self.train_steps: int = math.ceil(self.num_train_examples / self.args.train_batch_size)
 
-        ds = (
-            self.train_dataset.cache()
-            .shuffle(self.num_train_examples)
-            .batch(self.args.train_batch_size, drop_remainder=self.args.dataloader_drop_last)
-            .prefetch(tf.data.experimental.AUTOTUNE)
-        )
+        if self.args.is_tpu_available():
+            ds = (
+                self.train_dataset.cache()
+                .shuffle(self.num_train_examples)
+                .batch(self.args.train_batch_size * self.args.gradient_accumulation_steps, drop_remainder=self.args.dataloader_drop_last)
+                .unbatch()
+                .batch(self.args.train_batch_size)
+                .prefetch(tf.data.experimental.AUTOTUNE)
+            )
+        else:
+            ds = (
+                self.train_dataset.cache()
+                .shuffle(self.num_train_examples)
+                .batch(self.args.train_batch_size, drop_remainder=self.args.dataloader_drop_last)
+                .prefetch(tf.data.experimental.AUTOTUNE)
+            )
 
         if self.args.max_steps > 0:
             self.train_dataset = self.train_dataset.repeat(-1)
```

Comment:
@jplu Aha ! This is the error ! I found it !!

So as I mentioned in a previous comment, apparently we can't catch `OutOfRange` exception on TPU, which means we can't iterate the dataset epoch by epoch. Instead we should **ALWAYS** repeat the dataset, and iterate a specific number of steps. Same for validation / test dataset, as mentioned in the Kaggle notebook I linked previously. So there is quite a lot of change to make it work on TPU...

Anyway this line is wrong, it should be :
```python
ds = ds.repeat()
```

Reply:
Repeat is needed only if you run with `max_steps` > 1. But for the same reason `.repeat()` should be applied over the validation dataset indeed.

Reply:
In the case where the user specify `num_train_epochs` (`max_steps` = 0), the code rely on this try-catch to not iterate too far :

https://github.com/huggingface/transformers/blob/514410407019e9d20de7c2555595bab145c39ac0/src/transformers/trainer_tf.py#L385-L389

In my case, this try-catch is not working on TPU. So I need to rely on the behavior of `max_steps` > 1.

---

In the Kaggle notebook linked above, they mentioned that recreating a new iterator every epoch slow the training because TPU needs to re-compile the graph at the beginning of every epoch. Maybe we want to keep providing the `num_train_epochs` / `max_steps` behavior, but under the hood use only 1 way to iterate : `repeat()` with specific number of steps. 

What do you think ?

Reply:
You are right. What I have more in mind is: if someone wants to use TPUs he has to fullfil the `max_steps` parameter and keep like it is for GPUs/CPU.

The why of this try/catch is because it is more optimized for dataset iteration for CPU/GPUs.

Reply:
Later I will rethink the full process in order to be able to better handle the three cases (CPU/GPUs/TPUs).